test/diffstat_sub_async.go

```diff
@@ -0,0 +1,193 @@
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"regexp"
+	"sort"
+	"strings"
+)
+
+const NOISE_THRESHOLD = 0.03
+
+type Key struct {
+	Subs     int `json:"subs"`
+	Threads  int `json:"threads"`
+	Messages int `json:"messages"`
+}
+
+type TestData struct {
+	Key
+	Best    int `json:"best"`
+	Average int `json:"average"`
+	Worst   int `json:"worst"`
+}
+
+type Diff struct {
+	Key
+	BaseAverage   int     `json:"base"`
+	BranchAverage int     `json:"branch"`
+	Diff          float64 `json:"diff"`
+}
+
+type DiffData struct {
+	Records []Diff `json:"records"`
+
+	Total struct {
+		BestDiff    float64 `json:"best"`
+		AverageDiff float64 `json:"average"`
+		WorstDiff   float64 `json:"worst"`
+	} `json:"total"`
+}
+
+func main() {
```

Comment:
I have run the bench from this branch twice, to run this program to see if it worked ok. Even though it was the same code, I see quite a bit of variation between 2 runs:
```
== BenchSubscribeAsync_Large ==
Best: -0.05%
Average: 0.44%
Worst: 0.75%

== BenchSubscribeAsync_Small ==
Best: -4.10%
Average: 2.44%
Worst: 4.28%

== BenchSubscribeAsync_Inject ==
Best: 5.92%
Average: 3.45%
Worst: 2.60%

== BenchSubscribeAsync_InjectSlow ==
Best: 0.44%
Average: -0.03%
Worst: -1.34%
```
and was getting a lot of things like that:
``
subs=10 threads=7 messages=100000 base=48 branch=78 diff=62.50%
subs=10 threads=11 messages=100000 base=98 branch=148 diff=51.02%
subs=83 threads=7 messages=99932 base=50 branch=71 diff=42.00%
```
(I guess some that are excluded because too much difference?)

That is to say that I am not sure the validity of those bench comparisons, there seem to be too much variance even on the same branch.

Not sure what the right answer to that is. When I was testing changes, I would simply run the `./bin/nats-subscriber` and use the publisher and repeat that a bit and it was usually pretty stable. I would then repeat the test with my changes to see if the performance was affected. If the changes may affect the publisher but I wanted to check the subscriber, I would use a Go nats bench.
But that was "easy". Here we want to test possibly multiple subscriptions with/without they own dispatcher, etc..

Reply:
> noise.

The noise of the benchmark has been a big problem. The `inject` benchmark is a little less noisy, but the experience of the last couple weeks showed it to be insufficient.

The diff summarizer hides the diffs that under 5% (well within the noise for the individual tests) but also adds up all the _different_ keys and displays the (average) aggregate differences for the test. While not a great indicator I found it the best indicator of overall performance gain/loss.

What I'd like to do _after_ this series of PRs (once we are satisfied with the results) is to narrow down the matrix and re-run each remaining test enough time to detect/exclude the outliers (10+?). Then output each run individually, and use a go statistics package to summarize the results

> run the ./bin/nats-subscriber ... But that was "easy". Here we want to test possibly multiple subscriptions with/without they own dispatcher, etc...

Exactly. When I first remarked that I did some superficial perf analysis, that was on my MacOS M2 running the examples. It (almost) did not show a difference. With this benchmark many issues were found. For example, switching from `bool` to `:1` had no discernible difference on M2, but resulted in 2-3% performance drop in the "real IO" large/small benchmarks on x86 Ubuntu in CI. (I was quite surprised, BTW, but decided not to track it down, rolled back the change). Also, like I said, my original locking changes performed marginally better with large number of threads/subs, but underperformed with the main use-case of <10 subs.